### ISSUE TITLE
Update badge

### ISSUE THREAD

**Issue by fforootd** [caos/zitadel#151]
We should create a `update badge` in the admin section of `console`
More or less like gitlab does this (Screen Shot).

<img width="382" alt="Screenshot 2020-05-28 at 16 55 54" src="https://user-images.githubusercontent.com/9879976/83157768-4226c680-a104-11ea-9b17-8d4b017793b8.png">

**Comment by doncicuto** [comment]
Hi @fforootd and @mffap, how about this?

https://github.com/zitadel/zitadel/assets/30386061/7a81b126-b90e-4ebd-b007-7fe0d87016a0

In this video, I'm pretending that 2.30.0 is not the latest version, so the 'update asap'  message is shown. If you click on the badge you're redirected to Zitadel's releases pages, I'm not sure if I can implement a click to update action so as a workaround I think it's enough to point the user to Github.

My proposed code is developed in the frontend, if @peintnermax agrees. I fetch (HttpClient) Github's API URL: https://api.github.com/repos/zitadel/zitadel/releases/latest to know which is Zitadel's latest release and get its tag_name and html_url

I've thought about translating 'update asap' but asap is hard to translate so I personally would prefer to keep that badge untranslated.

Any thought, changes? 

Thanks!

**Comment by fforootd** [comment]
Nice work again, thank you!

I think we can keep this untranslated for the moment.

Just a thought that occurred to me while thinking about this.
There might be two "update states"

1) Update ASAP - when we need to advise people to pull an update fast due to security/performance reasons
2) Update - when people just should update

I had not yet a great idea how to solve this 😁 but I will give it a thought today
Maybe it is easier that we deploy a /update.json on our marketing page to read where we could read off versions with a tag (critical = true/false) or something similar.

**Comment by doncicuto** [comment]
Hi @fforootd,
I agree with your proposal with two states and if you could provide a update.json it would be even better as right now I'm requesting GitHub's API using its "free" tier and if you reach the requests limit it would affect how the badge work (low probability, of course). 

I'll wait till you have time to grasp the best solution for the update.json endpoint. 

Thanks!

**Comment by fforootd** [comment]
Ok I think I got an idea

We could do an API with zitadel.com/update?version={version}

The response can then be something like this:

```json
{
    "version": "v2.29.0",
    "critical": true,
    "message": "please update your app to the latest version",
    "url": "a url to a message we want to show the user"

}
````

**Comment by doncicuto** [comment]
Thanks @fforootd, your suggested json response looks perfect, let's see if I get how to use it.

As we're using a badge then we would query the API passing my current version and we have two possible messages for the badge: update (if critical is false) or update asap (if critical is true), right? Maybe we can set different background colors for these two states so ASAP is in red.

Now, if we click ok the badge we would be redirected to the URL set in json response, don't we? Showing the release note for example.

But the message that comes in the API response would be used un a tooltip for the badge, or where should we use it?

**Comment by fforootd** [comment]
> Thanks @fforootd, your suggested json response looks perfect, let's see if I get how to use it.
> 
> As we're using a badge then we would query the API passing my current version and we have two possible messages for the badge: update (if critical is false) or update asap (if critical is true), right? Maybe we can set different background colors for these two states so ASAP is in red.
> 
> Now, if we click ok the badge we would be redirected to the URL set in json response, don't we? Showing the release note for example.
> 
> But the message that comes in the API response would be used un a tooltip for the badge, or where should we use it?

Yeah you got this right, redirect should point to infos like release notes or security warnings and the message should be for a tooltip.

**Comment by doncicuto** [comment]
Cool, thanks @fforootd. I'll wait until the endpoint is ready as I can't code and endpoint for zitadel.com. Time to move to another issue 😄

**Comment by fforootd** [comment]
Hehe let me cook up something ASAP

**Comment by fforootd** [comment]
Hey I created a poc for this

I added an additional field "latest_version" for the gui to render.

https://[website-git-updates-zitadel.vercel.app/updates?version=v2.29.0](https://website-git-updates-zitadel.vercel.app/updates?version=v2.29.0)

It can react in three ways currently:

1. No version found (version pre 2.29.0 will not be found ATM)
2. No version provided
3. Version found

**Comment by doncicuto** [comment]
Hi @fforootd, thanks for the POC.

I've added the POC endpoint to my code. Here are two screenshots. The first one shows current situation: instance version 2.30 and endpoint response only suggests to update:

![Captura desde 2023-07-25 11-30-20](https://github.com/zitadel/zitadel/assets/30386061/4c20e8a8-27b2-4f5f-b6ab-8691129a47bb)

The second is simulating that 2.31.1 is a critical update:

![Captura desde 2023-07-25 11-48-24](https://github.com/zitadel/zitadel/assets/30386061/bc9fbc43-4d9d-4583-ab69-235e13eb9005)

The tooltip has been added with a default 'please update your app to the latest version' if no message is in the response. 

What do you think? I've added the suggested version to the badge but I may leave it to update or update asap

**Comment by fforootd** [comment]
Nice work.

You bring up a good point. I currently only have the latest version inside the response but not a `update to version >v2.31.1` or something similar. But looking at this such an information makes sense to me.

What do you think of this?

**Comment by doncicuto** [comment]
As we can see in this discussion https://gitlab.com/gitlab-org/gitlab/-/issues/344682#note_1233579698 alerting users is somewhat complicated but I like simple things.

I think we should keep the badge simple: "update" or "update asap", that's enough for a sensible administrator and more is useless to careless admins.

I'd prefer the API to tell me which is the latest version (latest_version), if I'm running a version that has critical security updates (critical) and which url has information about this critical situation (url that points me to zitadel.com information and how I should update). Those are covered with your proposed endpoint.

A more complete message in the response would be useful if, for example, we decide to include a warning when you log in the console and the API says we have critical situation!  But in that example (screenshot) how could we filter which users should see that message in the console? If we only use the message response for a tooltip for the badge then I don't think a complex message is needed.

![image](https://github.com/zitadel/zitadel/assets/30386061/e20f2570-a644-4da6-9a1f-30b48a42eaff)

As the badge is shown only when you visit your instance, we should warn somewhere else in the console if something critical affects us. I run Matomo for my blog analysis and it shows me a badge e when I do need to update as soon as I log in.

**Comment by doncicuto** [comment]
@fforootd this issue in crypto/tls (https://github.com/golang/go/issues/61460) would be a good example for this badge

**Comment by fforootd** [comment]
> @fforootd this issue in crypto/tls ([golang/go#61460](https://github.com/golang/go/issues/61460)) would be a good example for this badge

Yes, true that make absolut sense. Let me work out the last few things on the API 😁

**Comment by mffap** [comment]
> > @fforootd this issue in crypto/tls ([golang/go#61460](https://github.com/golang/go/issues/61460)) would be a good example for this badge
> 
> Yes, true that make absolut sense. Let me work out the last few things on the API 😁

@fforootd what's the state of this work?

**Comment by fforootd** [comment]
This feature does (not yet) have an ETA since I did not prioritize it to high.

What we have is a POC on both ends. The UI as well as an API.

I would recommend we move this to the regular sprint cycle and discuss priority.

cc @hifabienne

**Comment by hifabienne** [comment]
> This feature does (not yet) have an ETA since I did not prioritize it to high.
> 
> What we have is a POC on both ends. The UI as well as an API.
> 
> I would recommend we move this to the regular sprint cycle and discuss priority.
> 
> cc @hifabienne

From my perspective this feature is not really urgent and doesn't need high priorisation. Thats why it is in the backlog right now.
If you do see that differently I can take the necessary steps to estimate and plan the feature.

**Comment by mffap** [comment]
We should also consider showing a message in the logs, f.e. when an important update is due. Not sure what's better: showing this on startup only or also logging an info when we receive that information.
With that notifications could be automated by operators. Maybe there is also a more standardized way of doing this?

**Comment by fforootd** [comment]
> We should also consider showing a message in the logs, f.e. when an important update is due. Not sure what's better: showing this on startup only or also logging an info when we receive that information. With that notifications could be automated by operators. Maybe there is also a more standardized way of doing this?

Hm, yeah we could integrate an upgrade check and output a stdout message for example daily.
